### ISSUE TITLE
[$500] Improve Tab Navigation & Accessibility

### ISSUE THREAD

**Issue by kosmydel** [Expensify/App#36476]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** Latest
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Email or phone of affected tester (no customers):** N/A
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:** N/A
**Issue reported by:** me
**Slack conversation:** N/A

## Action Performed:
1. Open report.
2. Click the user avatar (next to the message).
3. The RHP opens.
4. Try to tab navigate.

## Expected Result:
You should navigate only inside the RHP modal. You shouldn't be able to escape it (unless you press escape/back button). This would improve the accessibility for users with visual disabilities.

## Actual Result:
You can navigate outside the RHP modal.

## Workaround:
N / A

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/104823336/5418229e-1be4-4fb3-b801-a52bbe7b33d2

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01681c9914c65b84bc</li>
        <li>Upwork Job ID: 1758337605760794624</li>
        <li>Last Price Increase: 2024-02-16</li>
<li>Automatic offers: </li>
<ul>
<li>c3024 | Reviewer | 0</li>
<li>fedirjh | Contributor | 0</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @isabelastisser</details>

**Comment by kosmydel** [comment]
Coming from [here](https://github.com/Expensify/App/issues/15631).
cc @roryabraham 

---

## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Users can navigate outside the RHP modal using the Tab/Shift+Tab.

### What is the root cause of that problem?
We are not using FocusTrap on the RHP modals.

### What changes do you think we should make in order to solve the problem?
Just add the focus trap like in [this PR](https://github.com/Expensify/App/pull/32800).

### What alternative solutions did you explore? (Optional)

N/A

**Comment by roryabraham** [comment]
Let's swap in @fedirjh as C+ since he's got context on the focus trap

**Comment by roryabraham** [comment]
Not overdue, go away melvin

**Comment by kosmydel** [comment]
Thanks @roryabraham!
I'm going to work on this next week, as this week I'm busy with higher priority issues and I'm OOO from Wednesday.

**Comment by trjExpensify** [comment]
Hey @kosmydel! @mountiny @puneetlath @iwiznia and I were talking internally about the inconsistency of keyboard/tab navigation in the app. We think it's time to tackle it holistically throughout the app, and we'll put it under the umbrella of the #vip-vsb project. Some places the keyboard/tab navigation works in the app, others it doesn't. We're forever getting "one-off" bugs related to it etc, some we fix and others we close waiting for a holistic approach. As a next step, @iwiznia had the idea to: 

1. Have one person implement tab navigation in one specific place (probably the expense creation flow as that is the most used flow I can think of that uses "forms")
2. Extract that work into something reusable (either a new component or a set of guidelines to follow for keyboard navigation)
3. Update the checklist to include that any new flow has to support keyboard navigation (by using the component/guidelines in 2)
4. Go back and fix all places where keyboard navigation was not implemented following the framework implemented above.

1-3 we'd be keen for you to take on if you want, given you're somewhat making a start by creating this issue. 4 could be as well if you want, but we could also just open that up to the community at large to fix the remaining bugs once 1-3 is in place.

What do you think?

**Comment by kosmydel** [comment]
@trjExpensify the plan sounds great!

Let me discuss it with my team internally, as I might not have enough capacity to handle it right now.

**Comment by trjExpensify** [comment]
Awesome, keep us posted!

**Comment by isabelastisser** [comment]
@kosmydel any updates here? Thanks!

**Comment by kosmydel** [comment]
> bump @kosmydel

Hey @isabelastisser, sorry for the delay. I've talked with my team in Software Mansion, and unfortunately, we don't have enough capacity to handle it this week. We should find someone to work on this next week, as it should be less busy.

**Comment by dragnoir** [comment]
Just for info, react-focus-trap is just a simple function that search for focusable elements htrow the page and loop the focus on those elements. We can copy the function, take what we need and enhance with hooks we already have.

**Comment by isabelastisser** [comment]
@roryabraham @trjExpensify I'm unsure of the next steps for this one.

**Comment by roryabraham** [comment]
we don't currently have an assignee actively working on this. I do not have any plans to push it forward myself.

So I guess we need to find another assignee, which according to @kosmydel should be sometime next week.

I do think that it would be good to clarify the prioritization of this issue, as I'm not sure it's gone through WN.

**Comment by roryabraham** [comment]
In my opinion weekly priority seems more appropriate here

**Comment by trjExpensify** [comment]
Yes, and I would deem it a `New feature` not a `Bug` really. As for priority, I think it would be a #vip-vsb project.

**Comment by isabelastisser** [comment]
Added to VIP-VSB.

**Comment by adamgrzybowski** [comment]
Hi! @filip-solecki and I are currently working on this issue

**Comment by isabelastisser** [comment]
Hey @adamgrzybowski and @filip-solecki, any updates? Thanks!

**Comment by adamgrzybowski** [comment]
We had some conversation [here](https://swmansion.slack.com/archives/C036QM0SLJK/p1710426496186039). It looks like some components related to syncing tab and arrow navigation are in refactor currently. I decided to focus my attention on focus traps. @kosmydel gave me some insights and it looks like there is quite a lot of that cases that we have to consider.

**Comment by isabelastisser** [comment]
Hey, @adamgrzybowski, I'm not sure what this means. Can you please clarify it? Thanks! cc @roryabraham

**Comment by roryabraham** [comment]
It's a bit vague, but for some context we previously had [this PR](https://github.com/Expensify/App/pull/32800), so it sounds like @adamgrzybowski is taking a similar approach to solve the problem described in the issue description.

**Comment by adamgrzybowski** [comment]
That's correct. I experimented a little and it looks like placing the focus trap higher in the hierarchy gives the best results. e.g. wrap the whole RHP Navigator. It limits mounting and unmounting the focus trap component.

Currently, I am looking into returning focus after unmounting the focus trap.

I assume that if we are going back to the screen with autofocus, then this element should be focused. If not then we should focus element focused before activating the focus trap. Is that right? @roryabraham

**Comment by adamgrzybowski** [comment]
Hi! I was working on few issues related to the improve tab navigation & accessibility and here is my quick summary

### Syncing multiple focus traps 
Initially, I tested using focus trap only for navigators to avoid problems with syncing but that wasn't ideal. When navigating back inside e.g. RHP the focus wouldn't be returned to the element that opened the screen. 

Instead, I went back to the idea of using focus-trap for the screen wrapper. There is a `trapStack` prop that helps syncing multiple focus traps. It seems to work well in my draft.

### Autofocus 
I was looking for a general solution to cover this requirement:

> I assume that if we are going back to the screen with autofocus, then this element should be focused. If not then we should focus element focused before activating the focus trap. Is that right? @roryabraham

I experimented with a solution where the focus trap waits for the screen to finish animating the transition to check if a text input is focused. In that case, the focus trap wouldn't set focus by itself. 

This solution looked promising and adding new screens with autofocus wouldn't require any work. But I couldn't make it work. There were always some broken cases.

After some consideration, it looks like the best solution is just to have a list with the screens having autofocus and avoid overriding focus there.

I created a draft with this solution [here](https://github.com/Expensify/App/pull/39520). It would be great if somebody could test it a little. I think I haven't listed all screens with autofocus and there may be more unique cases that we need to handle.

<details>
<summary>Video</summary>

https://github.com/Expensify/App/assets/67908363/03839179-c307-4748-a4a6-fb584116bc6d

</details>

### Wrong focus when going back to already opened screens.
This one is fixed in my draft. It turned out that some pressables were defocusing after an interaction. This was the reason why some screens had focused `document.body` on going back to them.


### Syncing tab and arrows
I passed my context to @WojtekBoman and it looks like he already posted a PR with a solution for this one 🚀 

## Remaining accessibility issues
The topic of accessibility turned out to be much broader than I initially expected. There are still some things that should be addressed. I am wondering if we could create separate issues for them so we could divide the work more easily. Especially given that I have a list of pending issues related to navigation that I should give my attention to.

### Confirming interaction with enter
It seems like confirming interaction with enter can trigger multiple calls of `onPressHandler` in the `BaseGenericPressable`. This can break the app sometimes e.g. confirming `goBack` with enter can pop too many screens from the stack.

### Screens with broken tab navigation
I found two screens where not all elements are accessible with tab navigation. Send money screen and request money screen. They need more investigation.

<details>
<summary>Screens</summary>

<img width="530" alt="image" src="https://github.com/Expensify/App/assets/67908363/a0572ce3-8a58-4d2f-9719-e2d4ed9f88d4">
<img width="530" alt="image" src="https://github.com/Expensify/App/assets/67908363/70a958aa-871d-4c41-8547-71b01d306141">

</details>

cc: @isabelastisser @roryabraham

**Comment by roryabraham** [comment]
> I am wondering if we could create separate issues for [accessibility issues] so we could divide the work more easily

this sounds like a great idea, but we might want to discuss prioritization this because accessibility isn't really the most critical concern for our roadmap today. Let's frame these accessibility issues as problem statements, then post them in slack to discuss which one(s) we want to pursue fixing right now. We probably want to prioritize the highest-ROI things first

**Comment by roryabraham** [comment]
overall thanks for your investigation, it looks great

**Comment by jnowakow** [comment]
Hello, I'll take over this issue from Adam as he's out of office now. 
@dragnoir I don't have so much insight as Adam but I'll try to start the conversation about the issues you mention above. 

> Right click on a reportItem, where focus should start and where focus should be when the BaseReportActionContextMenu will be removed.
I think that when the `BaseReportActionContextMenu` is open it should be the place when focus is trapped. Right now you can focus all elements on page which I don't think it good. 

<details>
<summary>Video</summary>


https://github.com/Expensify/App/assets/56261019/b97e5c5c-7e65-4e94-92b1-fe9527840f19


</details>

When the `BaseReportActionContextMenu` is dismissed then focus return to it previous state - nothing is focused. 


> In the context of using OptionsSelector, when one element is selected and another is focused using the tab key, what action should the enter key trigger? Your insights on this would be invaluable. Here's a video of the scenario for further examination:

I believe it was fixed by Wojtek [here](https://github.com/Expensify/App/pull/39201). Arrow and tab navigation is in sync and there situation that other entries are selected and focused at the same time.

**Comment by trjExpensify** [comment]
Assigning @jnowakow for tracking and moved back to `Weekly`. 👍

**Comment by mountiny** [comment]
This should be unblocked and you can resume work on the PR @jnowakow

**Comment by isabelastisser** [comment]
@jnowakow, please this Vit's comment. Thanks!

https://github.com/Expensify/App/issues/36476#issuecomment-2134143679

**Comment by techievivek** [comment]
@jnowakow, do you have an update on this issue? We have a few other issues on hold because of this one, so an update would be greatly appreciated. Thanks!

**Comment by jnowakow** [comment]
I'm not working on this one for quite a while. When it was merged I fixed some issues that appeared. Since then I assumed it was ready

**Comment by roryabraham** [comment]
We've added focus trap, which helps improve tab navigation in the RHP. It's a step in the right direction.

But I'm going to close this issue out. The original bug report is fixed, and the other kind of lofty goals of improving tab navigation across the app are not tightly scoped or high roadmap priorities.

**Comment by roryabraham** [comment]
sorry about that @fedirjh

**Comment by isabelastisser** [comment]
Payment summary:

@fedirjh $500 for C+ review. Pending via NewDot.

@fedirjh, can you please confirm this is correct? Thanks!

**Comment by JmillsExpensify** [comment]
$500 approved for @fedirjh

**Comment by isabelastisser** [comment]
All set. Closing!
